p2p/libp2p/module.go

```diff
@@ -0,0 +1,338 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	"github.com/libp2p/go-libp2p"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/pokt-network/pocket/logger"
+	"github.com/pokt-network/pocket/p2p/common"
+	"github.com/pokt-network/pocket/p2p/libp2p/identity"
+	"github.com/pokt-network/pocket/p2p/libp2p/protocol"
+	"github.com/pokt-network/pocket/p2p/providers"
+	"github.com/pokt-network/pocket/p2p/providers/addrbook_provider"
+	"github.com/pokt-network/pocket/p2p/providers/addrbook_provider/persistence"
+	"github.com/pokt-network/pocket/p2p/stdnetwork"
+	typesP2P "github.com/pokt-network/pocket/p2p/types"
+	"github.com/pokt-network/pocket/runtime/configs"
+	poktCrypto "github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/messaging"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+var _ modules.P2PModule = &libp2pModule{}
+
+type libp2pModule struct {
+	log                   *modules.Logger
+	bus                   modules.Bus
+	cfg                   *configs.P2PConfig
+	addrBookProvider      providers.AddrBookProvider
+	currentHeightProvider providers.CurrentHeightProvider
+	identity              libp2p.Option
+	listenAddrs           libp2p.Option
+	host                  host.Host
+	pubsub                *pubsub.PubSub
+	topic                 *pubsub.Topic
+	subscription          *pubsub.Subscription
+	network               typesP2P.Network
+}
+
+// NewReadStreamDeadline returns a future deadline
+// based on the read stream timeout duration.
+func NewReadStreamDeadline() time.Time {
+	return time.Now().Add(readStreamTimeoutDuration)
+}
+
+var (
+	// TODO: consider security exposure of and "safe minimum" for timeout.
+	// TODO: parameterize and expose via config.
+	// ReadStreamTimeout is the duration to wait for a read operation on a
+	// stream to complete, after which the stream is closed ("timed out").
+	readStreamTimeoutDuration = time.Second * 10
+	// ErrModule wraps errors which occur within the libp2pModule implementation.
+	ErrModule = common.NewErrFactory("LibP2P module error")
+	// ErrCloseStream wraps errors which occur when attempting to close a peer stream.
+	ErrCloseStream = common.NewErrFactory("an error occurred while closing peer stream")
+)
+
+func Create(bus modules.Bus) (modules.Module, error) {
+	return CreateWithProviders(
+		bus,
+		persistence.NewPersistenceAddrBookProvider(bus),
+		bus.GetConsensusModule(),
+	)
+}
+
+func CreateWithProviders(bus modules.Bus, addrBookProvider addrbook_provider.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+	return new(libp2pModule).CreateWithProviders(bus, addrBookProvider, currentHeightProvider)
+}
+
+func (mod *libp2pModule) GetModuleName() string {
+	// TODO: double check if this should change.
```

Comment:
Since it seems like you're implementing t as a completely different module, we should have a different name (until they're consolidated)

Reply:
My thinking was that the p2p modules are used mutually exclusively so I was expecting you to say it should be the same. Where does this present?

Reply:
To add some detail to my thinking here, I would expect that since these modules are both implementations of the `P2PModule` interface, that it wouldn't make sense to duplicate `shared/modules/p2p_module.go` as part of isolating the 2 p2p modules from one another (i.e. I would expect then to share that interface). It just so happens that `P2PModuleName` is defined in the same place and the way I read that was something like "this is the abstract label for that kind of module".

Reply:
Fair enough, let's remove the `// TODO: double check if this should change.`